pyaiot/gateway/common/common.py

```diff
@@ -0,0 +1,116 @@
+# Copyright 2017 IoT-Lab Team
+# Contributor(s) : see AUTHORS file
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice,
+# this list of conditions and the following disclaimer.
+#
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+#
+# 3. Neither the name of the copyright holder nor the names of its contributors
+# may be used to endorse or promote products derived from this software without
+# specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+"""Common classes for gateways."""
+
+import json
+import logging
+from tornado.ioloop import PeriodicCallback
+from tornado import web, gen
+from tornado.websocket import websocket_connect
+
+from pyaiot.common.auth import auth_token
+
+logger = logging.getLogger("pyaiot.gw.common")
+
+
+class GatewayBase(web.Application):
+    """Tornado based gateway application for CoAP nodes on a network."""
+
+    def __init__(self, keys, options):
+        if options.debug:
+            logger.setLevel(logging.DEBUG)
+
+        self.broker = None
+        self.keys = keys
+        handlers = []
+        settings = {'debug': True}
+
+        # Setup nodes controller
+        self._nodes_controller = self.setup_nodes_controller()
```

Comment:
Not completely sure on this, but maybe add a dummy `setup_nodes_controller()` to this class

Reply:
You hit the weak point of this PR :)
With this version, it can happen that a concrete gateway doesn't implement this method and the `_nodes_controller` could be `None`

I changed that to use multiple inheritance using `ABCMeta` and setting the `setup_nodes_controller` as pure virtual. Then all concrete gateway implementations have to implement this method. I tested it and it works well: an error message is displayed when the method is not implemented.
We could go even further by applying a similar design to the node controllers: a base class with abstract methods and concrete classes implementing them.

Reply:
Nice, this is exactly what I hoped to achieve with this comment.